src/software/ai/motion_constraint/motion_constraint_set_builder.cpp

```diff
@@ -0,0 +1,77 @@
+#include "motion_constraint_set_builder.h"
+
+std::set<MotionConstraint> MotionConstraintSetBuilder::buildMotionConstraints(
+    const GameState &game_state, Tactic &tactic)
+{
+    current_allowed_constraints.clear();
+    std::set<MotionConstraint> current_motion_constraints =
+        buildMotionConstraintsFromGameState(game_state);
+    try
+    {
+        // updates current_allowed_constraints
+        motion_constraint_visitor.set_current_allowed_constraint(
+            current_allowed_constraints);
+        tactic.accept(motion_constraint_visitor);
+    }
+    catch (std::invalid_argument)
+    {
+        // Tactic didn't implement accept so don't add any more motion constraints
+    }
+    for (const auto &constraint : current_allowed_constraints)
+    {
+        current_motion_constraints.erase(constraint);
+    }
+
+    return current_motion_constraints;
+}
+
+std::set<MotionConstraint>
+MotionConstraintSetBuilder::buildMotionConstraintsFromGameState(
+    const GameState &game_state) const
+{
+    std::set<MotionConstraint> motion_constraints;
+    motion_constraints.insert(MotionConstraint::FRIENDLY_DEFENSE_AREA);
+    motion_constraints.insert(MotionConstraint::ENEMY_ROBOTS_COLLISION);
+
+    if (game_state.stayAwayFromBall())
+    {
+        motion_constraints.insert(MotionConstraint::HALF_METER_AROUND_BALL);
+    }
+
+    if (game_state.isPenalty())
+    {
+        if (game_state.isOurPenalty())
+        {
+            motion_constraints.insert(MotionConstraint::ENEMY_HALF);
+        }
+        else
+        {
+            // Is their penalty
+            motion_constraints.insert(MotionConstraint::FRIENDLY_HALF);
+        }
+    }
```

Comment:
In this case the `game_state.isPenalty()` is essentially redundant as it makes no difference in the `isOurPentaly()` and the `else` branch

Reply:
I guess I'm just kinda confused about the `else` case for the largest scope `if`

Reply:
I have created another issue with respect to your comments about the logic in the motion_constraint_set_builder since it is outside the scope of this PR: https://github.com/UBC-Thunderbots/Software/issues/1821 . I will work on the change after this PR gets merged